iocore/utils/PropertyBlock.h

```diff
@@ -0,0 +1,254 @@
+#pragma once
+#ifndef _PropertyBlock_h
+#define _PropertyBlock_h
+
+#include "stdint.h"
+#include <cstddef>
+#include <atomic>
+#include <cstring>
+#include <vector>
+#include <type_traits>
+
+#include "ts/ink_assert.h"
+
+/// runtime defined static container
+/** The size of this structure is actually zero, so it will not change the size of your derived class.
+ * But new and delete are overriden to use allocate enough bytes of the derived type + properties added
+ * at run time using ioBufferAllocator.
+ *
+ * Also all bools are packed to save space using the *Bit methods.
+ *
+ * This is templated so static variables are instanced per Derived type. B/c we need to have different
+ * size property sets.
+ */
+template <class Derived_t> ///< static members instance pey type.
+struct PropertyBlock {
+public:
+  using self_t = PropertyBlock<Derived_t>;
+  /// callback format used to PropBlock init and destroy
+  using PropertyFunc_t = void(Derived_t *, void *);
+
+  /// number of bytes from 'this', or the bitset index
+  using Offset_t = uint32_t;
+
+  /// Add a data type to schema.
+  /**
+   * @param prop_count the number of instances of this data type to allocate.
+   * @param init an initialization function for property
+   * @param destroy a destruction function for property
+   * Note: if prop_count is 0, the init and destroy are still called per instance.
+   */
+  template <typename Property_t>
+  static Offset_t
+  PropBlockDeclare(size_t prop_count, PropertyFunc_t *init, PropertyFunc_t *destroy)
```

Comment:
It might be better to put `prop_count` last so you can default it to `1`, the common case.

Reply:
Should I just remove this parameter, and force the caller to write a loop?
calling PropBlockDeclare<int>(5) is exactly the same as for (int i=5;--;) PropBlockDeclare<int>(1)